Link flags and debug flags for Fujitsu compiler was not used properly

I'm trying to build ESMF 8.6.1 and 8.8.1 with Fujitsu compiler on A64FX platform; the compiler version is:

> fcc: Fujitsu C/C++ Compiler 4.11.2 tcsds-1.2.40
> simulating gcc version 6.1

The native compiler name for each language is:
C: (mpi)fcc
CXX: (mpi)FCC
Fortran: (mpi)frt

According to the document, when linking mixed-language (C/C++ and Fortran) objects, (mpi)frt should use:
--linkstl=libstdc++
and (mpi)fcc/FCC should use:
--linkfortran

The flags seems set correctly in build_config/Linux.fujitsu.default/build_rules.mk:
```
ESMF_F90LINKLIBS += --linkstl=libstdc++
ESMF_CXXLINKLIBS += --linkfortran
```
But when doing "gmake install", the following error was produced:
```
mpifcc -Nclang     -Kopenmp -L/nwpr/gfs/com132/fx1000/esmf-8.8.1/lib/libO/Linux.fujitsu.64.mpi.default -L/nwpr/gfs/com132/fx1000
/netcdf-4.9.2/lib -L/nwpr/gfs/com132/fx1000/netcdf-4.9.2/lib -L/nwpr/gfs/com132/fx1000/netcdf-4.9.2/lib -L/nwpr/gfs/com132/fx100
0/pnetcdf-1.14.0/lib
(......) -lhdf5_hl -lhdf5 -lz -lxml2 -lm -lpnetcdf -lpiof -lpioc -lpnetcdf --linkstl=libstdc++
clang-7: error: unsupported option '--linkstl=libstdc++'
make[4]: *** [/IFS/nwpr/gfs/com132/bld/esmf-8.8.1/build/common.mk:2711: /nwpr/gfs/com132/fx1000/esmf-8.8.1/bin/binO/Linux.fujitsu.64.mpi.default/ESMF_PrintInfoC] Error 1

```
The linker in this case is of C (mpifcc), but it uses the linker flag for Fortran compiler (--linkstl=libstdc++), thus gives an error.

Also, the debug flags in build_rules.mk uses:
`-Nquickdbg`
But, although not mentioned in the document, our collaborating Fujitsu engineer clarifies that this flag is not available in Clang mode; the following flags should be used in Clang mode:
`-fsanitize=address -fsanitize=undefined`
This does not affect Fortran compiler;  only C/C++ compiler has the Clang mode.

Mike

The pull works as expected, but now the mpifcc link gives whole bunch of undefined reference:
```
/nwpr/gfs/com132/fx1000/esmf-8.6.1-debug/lib/libg/Linux.fujitsu.64.mpi.default/lt/libesmf.a(ESMC_VM.o): In function 'ESMC VMLogMemInfo':
/IFS/nwpr/gfs/com132/bld/esmf-8.6.1/src/Infrastructure/W/interface/ESMC VM.C:286: undefined reference to 'std::allocatorkchar>::allocator( )'
```
Guess we'll need a -listdc++ for that; I'm testing this now.

--- a/build/common.mk
+++ b/build/common.mk
@@ -1134,7 +1134,8 @@ endif
 
 ifeq ($(ESMF_COMPILER),fujitsu)
 # under Fujitsu the Fortran link libs are not compatible with C linker
-ESMF_CLINKLIBS     += $(ESMF_CXXLINKLIBS)
+# explicitly set here
+ESMF_CLINKLIBS     += $(ESMF_CXXLINKLIBS) -lstdc++
 else
 # but for other compilers the Fortran link libs are often needed by C linker
 ESMF_CLINKLIBS     += $(ESMF_CXXLINKLIBS) $(ESMF_F90LINKLIBS)

--- a/build/common.mk
+++ b/build/common.mk
@@ -1131,7 +1131,14 @@ endif
 ifeq ($(origin ESMF_CLINKLIBS_ENV), environment)
 ESMF_CLINKLIBS = $(ESMF_CLINKLIBS_ENV)
 endif
+
+ifeq ($(ESMF_COMPILER),fujitsu)
+# under Fujitsu the Fortran link libs are not compatible with C linker
+ESMF_CLINKLIBS     += $(ESMF_CXXLINKLIBS)
+else
+# but for other compilers the Fortran link libs are often needed by C linker
 ESMF_CLINKLIBS     += $(ESMF_CXXLINKLIBS) $(ESMF_F90LINKLIBS)
+endif
 ESMF_CESMFLINKLIBS += -lesmf $(ESMF_CLINKLIBS)
 
 # - tools: AR + RANLIB + ...

--- a/build_config/Linux.fujitsu.default/build_rules.mk
+++ b/build_config/Linux.fujitsu.default/build_rules.mk
@@ -61,8 +61,8 @@ ESMF_CXXCOMPILECPPFLAGS += -DESMF_NO_F2018ASSUMEDTYPE
 ############################################################
 # Special debug flags
 #
-ESMF_F90OPTFLAG_G       += -Nquickdbg 
-ESMF_CXXOPTFLAG_G       += -Nquickdbg
+ESMF_F90OPTFLAG_G       += -Nquickdbg
+ESMF_CXXOPTFLAG_G       +=
 
 ############################################################
 # Enable TR15581/F2003 Allocatable array resizing
@@ -133,7 +133,7 @@ ESMF_F90LINKLIBS += --linkstl=libstdc++
 ESMF_CXXLINKLIBS += --linkfortran
 
 ############################################################
-# Linker option that ensures that the specified libraries are 
+# Linker option that ensures that the specified libraries are
 # used to also resolve symbols needed by other libraries.
 #
 ESMF_F90LINKOPTS          += -Wl,--no-as-needed

